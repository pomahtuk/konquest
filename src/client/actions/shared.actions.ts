export enum SharedActionTypes {
  RESTORE_GAME = "RESTORE_GAME"
}

export interface RestoreGameAction {
  type: typeof SharedActionTypes.RESTORE_GAME;
  storedGame?: string;
}

export function restoreGame(storedGame?: string): RestoreGameAction {
  return { type: SharedActionTypes.RESTORE_GAME, storedGame };
}
