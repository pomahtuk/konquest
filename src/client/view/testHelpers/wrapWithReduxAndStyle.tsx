import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import storeCreator from "../../stores/game.store";
import ThemeProvider from "../themes/ThemeProvider";
import baseTheme from "../themes/base.theme";
const store = storeCreator();

export const CurrentStore = store;

const wrapWithReduxAndStyle = (element: ReactElement): ReactElement => {
  return (
    <Provider store={store}>
      <ThemeProvider theme={baseTheme}>{element}</ThemeProvider>
    </Provider>
  );
};

export default wrapWithReduxAndStyle;
