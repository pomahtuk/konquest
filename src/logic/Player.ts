export default class Player {
  public id: number;
  public screenName: string;

  public constructor(name: string) {
    // TODO: player actually have to be stored somewhere
    this.id = Math.floor(Math.random() * 1000);
    this.screenName = name;
  }
}

export interface PlayerMap {
  [key: string]: Player;
}
