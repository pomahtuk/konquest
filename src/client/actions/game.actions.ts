import Player, { PlayerTurn } from "../../logic/Player";

/*
 * action types
 */
export enum GameActionTypes {
  START_GAME = "START_GAME",
  SET_GAME_OPTIONS = "SET_GAME_OPTIONS",
  ADD_PLAYER_TURN = "ADD_PLAYER_TURN"
}

export interface StateGameOptions {
  fieldSize: number;
  neutralPlanetCount: number;
  players: Player[];
}

export interface StartGameAction {
  type: typeof GameActionTypes.START_GAME;
}

export interface SetGameOptionsAction {
  type: typeof GameActionTypes.SET_GAME_OPTIONS;
  gameOptions: StateGameOptions;
}

export interface AddPlayerTurnAction {
  type: typeof GameActionTypes.ADD_PLAYER_TURN;
  turnData: PlayerTurn;
}
/*
 * action creators
 */
export function startGame(): StartGameAction {
  return { type: GameActionTypes.START_GAME };
}

export function setGameOptions(gameOptions: StateGameOptions): SetGameOptionsAction {
  return { type: GameActionTypes.SET_GAME_OPTIONS, gameOptions };
}

export function addPlayerTurn(turnData: PlayerTurn): AddPlayerTurnAction {
  return { type: GameActionTypes.ADD_PLAYER_TURN, turnData };
}
