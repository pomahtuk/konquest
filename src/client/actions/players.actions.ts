import Player from "../../logic/Player";

export enum PlayerActionTypes {
  ADD_PLAYER = "ADD_PLAYER",
  REMOVE_PLAYER = "REMOVE_PLAYER",
  CLEAR_PLAYERS = "CLEAR_PLAYERS"
}

export interface AddPlayerAction {
  type: typeof PlayerActionTypes.ADD_PLAYER;
  player: Player;
}

export interface RemovePlayerAction {
  type: typeof PlayerActionTypes.REMOVE_PLAYER;
  player: Player;
}

export interface ClearPlayersAction {
  type: typeof PlayerActionTypes.CLEAR_PLAYERS;
}

// action creators
export function addPlayer(player: Player): AddPlayerAction {
  return { type: PlayerActionTypes.ADD_PLAYER, player };
}

export function removePlayer(player: Player): RemovePlayerAction {
  return { type: PlayerActionTypes.REMOVE_PLAYER, player };
}

export function clearPlayers(): ClearPlayersAction {
  return { type: PlayerActionTypes.CLEAR_PLAYERS };
}
