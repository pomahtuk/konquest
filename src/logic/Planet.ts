import Player from "./Player";

export interface PlanetCoordinates {
  x: number;
  y: number;
}

export default class Planet {
  public name: string;
  public owner: number | null = null; // reference co playerId
  public ships: number;
  public shipsDue: number | 0 = 0; // only if there are ships en route
  public production: number; // only happens when planet is taken, next turn
  public killPercent: number;
  // Kill percent is a measure of the effectiveness of the ships produced at that planet.
  // Attack fleets take the kill percentage of their planet of departure,
  // and defense fleets use the kill percentage of the planet they are defending.
  public coordinates: PlanetCoordinates;

  public constructor(name: string, player?: Player | null, coordinates: PlanetCoordinates = { x: 0, y: 0 }) {
    // initialize
    this.name = name;
    this.coordinates = coordinates;

    let production = 10;
    const killPercent = 0.5;
    if (!player) {
      production = 8 + Math.floor(Math.random() * 8);
      // TODO: generate killPercent
      this.ships = 0;
    } else {
      this.owner = player.id;
      this.ships = production;
    }

    this.production = production;
    this.killPercent = killPercent;
  }
}

export interface PlanetMap {
  [key: string]: Planet;
}
