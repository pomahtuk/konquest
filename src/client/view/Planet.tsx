import React, { ReactElement } from "react";
import { shallowEqual, useSelector, useDispatch } from "react-redux";
import { Tooltip } from "react-tippy";

import { setDestinationPlanet, setOriginPlanet } from "../actions/game.actions";

import Planet from "../../logic/Planet";

import "react-tippy/dist/tippy.css";
import planetSelectorFunction, { PlanetStoreSlice } from "../selectors/planet.selector";
import RandomImage from "./RandomImage";

export interface PlanetProps {
  planet: Planet;
}

const PlanetElement = ({ planet }: PlanetProps): ReactElement => {
  const { originPlanet, activePlayer, destinationPlanet, currentShipsModifier }: PlanetStoreSlice = useSelector(planetSelectorFunction, shallowEqual);
  const dispatch = useDispatch();

  if (!planet) {
    return <span>Â·</span>;
  }

  const onPlanetSelect = (): void => {
    if (!originPlanet && (!planet.owner || planet.owner.id !== activePlayer.id)) {
      // origin should belong to player
      return;
    } else if (!originPlanet && planet.owner && planet.owner.id === activePlayer.id) {
      dispatch(setOriginPlanet(planet));
    } else if (originPlanet && !destinationPlanet) {
      dispatch(setDestinationPlanet(planet));
    } else if (planet.owner && planet.owner.id === activePlayer.id) {
      dispatch(setOriginPlanet(planet));
      dispatch(setDestinationPlanet(undefined));
    }
  };

  return (
    <Tooltip
      position="top"
      animation="fade"
      arrow="true"
      animateFill="false"
      html={
        <div>
          <div>Owner: {planet.owner ? planet.owner.screenName : "none"}</div>
          <div>Production: {planet.production}</div>
          <div>Kill percent: {planet.killPercent}</div>
          <div>Ships: {planet.ships - (currentShipsModifier[planet.name] || 0)}</div>
        </div>
      }
      style={{
        position: "absolute",
        top: planet.coordinates.x * 60 - 10,
        left: planet.coordinates.y * 60 - 10,
        width: 60,
        height: 60
      }}
    >
      <div
        style={{
          cursor: "pointer",
          position: "relative",
          top: 10,
          left: 10,
          width: 40,
          height: 40
        }}
        onClick={onPlanetSelect}
      >
        <RandomImage imageKey={planet.name}></RandomImage>
        <span
          style={{
            position: "absolute",
            bottom: 0,
            right: 0,
            zIndex: 999
          }}
        >
          {planet.name}
        </span>
        <br />
      </div>
    </Tooltip>
  );
};

export default PlanetElement;
