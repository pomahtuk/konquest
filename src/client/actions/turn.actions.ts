import { PlayerTurnOrder } from "../../logic/Player";
import Planet from "../../logic/Planet";

/*
 * action types
 */
export enum TurnActionTypes {
  SET_ORIGIN_PLANET = "SET_ORIGIN_PLANET",
  SET_DESTINATION_PLANET = "SET_DESTINATION_PLANET",
  ADD_PLAYER_TURN_ORDER = "ADD_PLAYER_TURN_ORDER",
  REMOVE_PLAYER_TURN_ORDER = "REMOVE_PLAYER_TURN_ORDER",
  CLEAR_TURN_DATA = "CLEAR_TURN_DATA",
  UNSET_PLANETS = "UNSET_PLANETS"
}

export interface SetPlanetAction {
  type: typeof TurnActionTypes.SET_ORIGIN_PLANET | typeof TurnActionTypes.SET_DESTINATION_PLANET;
  planet?: Planet;
}

export interface AddPlayerTurnOrderAction {
  type: typeof TurnActionTypes.ADD_PLAYER_TURN_ORDER;
  order: PlayerTurnOrder;
}

export interface ClearTurnDataAction {
  type: typeof TurnActionTypes.CLEAR_TURN_DATA;
}

export interface RemovePlayerTurnOrderAction {
  type: typeof TurnActionTypes.REMOVE_PLAYER_TURN_ORDER;
  index: number;
}

export interface UnsetPlanetsAction {
  type: typeof TurnActionTypes.UNSET_PLANETS;
}

/*
 * action creators
 */
export function setOriginPlanet(planet?: Planet): SetPlanetAction {
  return { type: TurnActionTypes.SET_ORIGIN_PLANET, planet };
}

export function setDestinationPlanet(planet?: Planet): SetPlanetAction {
  return { type: TurnActionTypes.SET_DESTINATION_PLANET, planet };
}

export function addPlayerTurnOrder(order: PlayerTurnOrder): AddPlayerTurnOrderAction {
  return { type: TurnActionTypes.ADD_PLAYER_TURN_ORDER, order };
}

export function removePlayerTurnOrder(index: number): RemovePlayerTurnOrderAction {
  return { type: TurnActionTypes.REMOVE_PLAYER_TURN_ORDER, index };
}

export function clearTurnData(): ClearTurnDataAction {
  return { type: TurnActionTypes.CLEAR_TURN_DATA };
}

export function unsetPlanets(): UnsetPlanetsAction {
  return { type: TurnActionTypes.UNSET_PLANETS };
}
