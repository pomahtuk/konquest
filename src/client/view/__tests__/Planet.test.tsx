import React from "react";
import { render, fireEvent } from "@testing-library/react";

import wrapWithReduxAndStyle, { CurrentStore } from "../testHelpers/wrapWithReduxAndStyle";
import Planet from "../Planet";
import { setGameOptions, startGame } from "../../actions/game.actions";
import Player from "../../../logic/Player";

const testCoordinates = { x: 0, y: 0 };

describe("<Planet />", (): void => {
  beforeAll((): void => {
    CurrentStore.dispatch(
      setGameOptions({
        fieldSize: 4,
        neutralPlanetCount: 1,
        players: [new Player("1"), new Player("2")]
      })
    );
    CurrentStore.dispatch(startGame());
  });

  it("Renders planet not belonging to player without exploding when nothing selected and not allowing click", (): void => {
    const {
      game: { planets }
    } = CurrentStore.getState();
    // make sure tests would not fail due to random factors
    planets.B.mainColor = "#000";
    planets.B.coordinates = testCoordinates;
    const { container, getByTestId } = render(wrapWithReduxAndStyle(<Planet blockSize={40} planet={planets["B"]} />));
    expect(container).toMatchSnapshot();
    const planetElement = getByTestId("planet");
    fireEvent.click(planetElement);
    expect(CurrentStore.getState().turn.destinationPlanet).not.toBeDefined();
    expect(CurrentStore.getState().turn.originPlanet).not.toBeDefined();
  });

  it("Renders player planet without exploding when nothing selected", (): void => {
    const {
      game: { planets }
    } = CurrentStore.getState();
    // make sure tests would not fail due to random factors
    planets.A.mainColor = "#000";
    planets.A.coordinates = testCoordinates;
    const { container, getByTestId } = render(wrapWithReduxAndStyle(<Planet blockSize={40} planet={planets["A"]} />));
    expect(container).toMatchSnapshot();
    const planetElement = getByTestId("planet");
    fireEvent.click(planetElement);
    expect(CurrentStore.getState().turn.originPlanet.name).toBe("A");
  });

  it("Renders player planet without exploding when it is origin planet", (): void => {
    const {
      game: { planets }
    } = CurrentStore.getState();
    // make sure tests would not fail due to random factors
    planets.A.mainColor = "#000";
    planets.A.coordinates = testCoordinates;
    const { container } = render(wrapWithReduxAndStyle(<Planet blockSize={40} planet={planets["A"]} />));
    expect(container).toMatchSnapshot();
  });

  it("Renders neutral planet without exploding when origin selected and allowing to set this planet as destination", (): void => {
    const {
      game: { planets }
    } = CurrentStore.getState();
    // make sure tests would not fail due to random factors
    planets.C.mainColor = "#000";
    planets.C.coordinates = testCoordinates;
    const { container, getByTestId } = render(wrapWithReduxAndStyle(<Planet blockSize={40} planet={planets["C"]} />));
    expect(container).toMatchSnapshot();
    const planetElement = getByTestId("planet");
    fireEvent.click(planetElement);
    expect(CurrentStore.getState().turn.destinationPlanet.name).toBe("C");
  });

  it("Renders player planet without exploding when it is origin planet and deselecting all on click", (): void => {
    const {
      game: { planets }
    } = CurrentStore.getState();
    // make sure tests would not fail due to random factors
    planets.A.mainColor = "#000";
    planets.A.coordinates = testCoordinates;
    const { getByTestId } = render(wrapWithReduxAndStyle(<Planet blockSize={40} planet={planets["A"]} />));
    const planetElement = getByTestId("planet");
    fireEvent.click(planetElement);
    expect(CurrentStore.getState().turn.destinationPlanet).not.toBeDefined();
    expect(CurrentStore.getState().turn.originPlanet.name).toBe("A");
  });
});
