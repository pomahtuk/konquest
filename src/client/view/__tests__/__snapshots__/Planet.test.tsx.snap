// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Planet /> Renders neutral planet without exploding when origin selected and allowing to set this planet as destination 1`] = `
.emotion-1 {
  position: absolute;
  top: 6px;
  left: 6px;
  width: 28px;
  height: 28px;
  border-radius: 40px;
}

.emotion-0 {
  position: absolute;
  bottom: 0px;
  right: 0px;
  z-index: 999;
}

.emotion-2 {
  position: absolute;
  cursor: pointer;
  top: -1px;
  left: 0px;
  width: 39px;
  height: 39px;
  background: none;
}

.emotion-2:hover {
  background: rgba(0,0,255,0.3);
}

<div>
  <div
    class="emotion-2"
    data-testid="planet"
  >
    <div
      aria-describedby="tippy-tooltip-4"
      class="emotion-1"
      data-original-title="undefined"
      data-tooltipped=""
      style="display: inline;"
    >
      <svg
        height="100%"
        version="1.1"
        width="100%"
        xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink"
      >
        <defs>
          <lineargradient
            id="main_shadow-#000"
          >
            <stop
              offset="0"
              stop-color="#ccc"
            />
            <stop
              offset="0.6"
              stop-color="#000"
            />
          </lineargradient>
        </defs>
        <g
          transform="rotate(60, 14 14)"
        >
          <circle
            cx="14"
            cy="14"
            fill="url(#main_shadow-#000)"
            r="14"
          />
        </g>
      </svg>
      <span
        class="emotion-0"
      >
        C
      </span>
    </div>
  </div>
</div>
`;

exports[`<Planet /> Renders planet not belonging to player without exploding when nothing selected and not allowing click 1`] = `
.emotion-2 {
  position: absolute;
  cursor: pointer;
  top: -1px;
  left: 0px;
  width: 39px;
  height: 39px;
  background: none;
}

.emotion-2:hover {
  background: rgba(255,255,255,0.3);
}

.emotion-1 {
  position: absolute;
  top: 6px;
  left: 6px;
  width: 28px;
  height: 28px;
  border-radius: 40px;
}

.emotion-0 {
  position: absolute;
  bottom: 0px;
  right: 0px;
  z-index: 999;
}

<div>
  <div
    class="emotion-2"
    data-testid="planet"
  >
    <div
      aria-describedby="tippy-tooltip-1"
      class="emotion-1"
      data-original-title="undefined"
      data-tooltipped=""
      style="display: inline;"
    >
      <svg
        height="100%"
        version="1.1"
        width="100%"
        xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink"
      >
        <defs>
          <lineargradient
            id="main_shadow-#000"
          >
            <stop
              offset="0"
              stop-color="#ccc"
            />
            <stop
              offset="0.6"
              stop-color="#000"
            />
          </lineargradient>
        </defs>
        <g
          transform="rotate(60, 14 14)"
        >
          <circle
            cx="14"
            cy="14"
            fill="url(#main_shadow-#000)"
            r="14"
          />
        </g>
      </svg>
      <span
        class="emotion-0"
      >
        B
      </span>
    </div>
  </div>
</div>
`;

exports[`<Planet /> Renders player planet without exploding when it is origin planet 1`] = `
.emotion-1 {
  position: absolute;
  top: 6px;
  left: 6px;
  width: 28px;
  height: 28px;
  border-radius: 40px;
}

.emotion-0 {
  position: absolute;
  bottom: 0px;
  right: 0px;
  z-index: 999;
}

.emotion-2 {
  position: absolute;
  cursor: pointer;
  top: -1px;
  left: 0px;
  width: 39px;
  height: 39px;
  background: rgba(0,255,0,0.3);
}

.emotion-2:hover {
  background: rgba(0,255,0,0.3);
}

<div>
  <div
    class="emotion-2"
    data-testid="planet"
  >
    <div
      aria-describedby="tippy-tooltip-3"
      class="emotion-1"
      data-original-title="undefined"
      data-tooltipped=""
      style="display: inline;"
    >
      <svg
        height="100%"
        version="1.1"
        width="100%"
        xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink"
      >
        <defs>
          <lineargradient
            id="main_shadow-#000"
          >
            <stop
              offset="0"
              stop-color="#ccc"
            />
            <stop
              offset="0.6"
              stop-color="#000"
            />
          </lineargradient>
        </defs>
        <g
          transform="rotate(60, 14 14)"
        >
          <circle
            cx="14"
            cy="14"
            fill="url(#main_shadow-#000)"
            r="14"
          />
        </g>
      </svg>
      <span
        class="emotion-0"
      >
        A
      </span>
    </div>
  </div>
</div>
`;

exports[`<Planet /> Renders player planet without exploding when nothing selected 1`] = `
.emotion-1 {
  position: absolute;
  top: 6px;
  left: 6px;
  width: 28px;
  height: 28px;
  border-radius: 40px;
}

.emotion-0 {
  position: absolute;
  bottom: 0px;
  right: 0px;
  z-index: 999;
}

.emotion-2 {
  position: absolute;
  cursor: pointer;
  top: -1px;
  left: 0px;
  width: 39px;
  height: 39px;
  background: rgba(255,0,0,0.3);
}

.emotion-2:hover {
  background: rgba(255,0,0,0.3);
}

<div>
  <div
    class="emotion-2"
    data-testid="planet"
  >
    <div
      aria-describedby="tippy-tooltip-2"
      class="emotion-1"
      data-original-title="undefined"
      data-tooltipped=""
      style="display: inline;"
    >
      <svg
        height="100%"
        version="1.1"
        width="100%"
        xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink"
      >
        <defs>
          <lineargradient
            id="main_shadow-#000"
          >
            <stop
              offset="0"
              stop-color="#ccc"
            />
            <stop
              offset="0.6"
              stop-color="#000"
            />
          </lineargradient>
        </defs>
        <g
          transform="rotate(60, 14 14)"
        >
          <circle
            cx="14"
            cy="14"
            fill="url(#main_shadow-#000)"
            r="14"
          />
        </g>
      </svg>
      <span
        class="emotion-0"
      >
        A
      </span>
    </div>
  </div>
</div>
`;
